markdown: Handle prettified links in URL placeholder.

Harshbansal8705 opened the pull request:
Handles the stream/topic/message link when a prettified link is pasted in the URL part of a hyperlink text, such as: [text](#**channel**).

This implementation extracts prettified links prefixed with '](' and replaces them with plain text links, which are then processed by the native LinkInlineProcessor.

<!-- Describe your pull request here.-->

Fixes: #31985

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

|Before| After |
|---------|---------|
|![Screenshot_20250119_040301](https://github.com/user-attachments/assets/5dd213fc-d6bd-459b-8c5a-cf4f013b8d0c)|![Screenshot_20250119_040421](https://github.com/user-attachments/assets/c573c618-2139-4489-b927-da72e9a04afa)|

[Screencast_20250119_031537.webm](https://github.com/user-attachments/assets/b6cc8973-e967-430f-b2a3-d12fecfe5825)

I'm not very sure of the functions and variables I've named here. Any suggestion for improvement would be great.


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Harshbansal8705 committed:
markdown: Handle prettified links in URL placeholder.

Handles the stream/topic/message link when a prettified link is pasted
in the URL part of a hyperlink text, such as: [text](#**channel**).

This implementation extracts prettified links prefixed with '](' and
replaces them with plain text links, which are then processed by the
native LinkInlineProcessor.

Not implementing this regex in the frontend markdown processor since
it's an extremely rare corner case where a prettified link appears in
the URL placeholder of a markdown syntax. Also, the broken link is only
shows for a short duration during local echo before being corrected by
the backend. The complexity of implementing this in the web app isn't
worth it for such a rare case.

Fixes: #31985

alya commented:
@apoorvapendse or @kuv2707 , would either of you have a chance to review this one?

apoorvapendse reviewed (commented):
Thanks @alya, this works for me in manual testing.
Left a couple comments and a [doubt](https://github.com/zulip/zulip/pull/33086#discussion_r1924603496) for @kuv2707.

kuv2707 commented on zerver/lib/markdown/__init__.py:
```diff
@@ -167,18 +171,25 @@ def get_compiled_stream_link_regex() -> Pattern[str]:
+                     \#\*\*                                 # Hash sign followed by double asterisks
+                         (?P<stream_name>[^\*>]+)           # stream name can contain anything except >
+                         >                                  # > acts as separator
+                         (?P<topic_name>[^\*]+)             # topic name can contain anything
```
I have two arguments:

- *In favor*: The frontend and the backend markdown processors should be as consistent with each other as possible, so we should align this with the frontend. Though there will always be inconsistencies (see #32997)

- *Against*:  This is too rare a corner case to make so many changes for, and both the broken link and the correct one render identically. Moreover, the backend markdown processor is the ultimate source of truth, and the broken link rendered by the frontend markdown processor will only be there for a very small duration. (And on a slow connection, there are various other areas of the website that would get messed up before this).


Me personally, I lean a bit towards the second argument. 
@Harshbansal8705 Please ask about this in CZO.

---

If we do decide to implement this in the frontend, I leave for you an observation:
Unlike the backend processor, the frontend one does leave the `**` intace, making it much easier to convert it to a valid link using functions in `topic_link_util.ts` etc.

Typing this:
<img width="302" alt="image" src="https://github.com/user-attachments/assets/e0390652-dfed-4d19-b0df-339b08b41b52" />

Gives this:

<img width="512" alt="image" src="https://github.com/user-attachments/assets/8e5699a3-ada0-464c-94e5-f8a2a02fe3c1" />

kuv2707 commented on zerver/lib/markdown/__init__.py:
```diff
@@ -167,18 +171,25 @@ def get_compiled_stream_link_regex() -> Pattern[str]:
 
+BASE_STREAM_TOPIC_LINK_REGEX = r"""
+                     \#\*\*                                 # Hash sign followed by double asterisks
+                         (?P<stream_name>[^\*>]+)           # stream name can contain anything except >
```
I think changes similar to [this](https://github.com/zulip/zulip/pull/33086/files#r1924602800) in the comments should be done in other places too.

kuv2707 reviewed (commented):
Left some comments.

Harshbansal8705 commented:
#### Updates:
- Updated the comments as suggested:
  > I think changes similar to [this](https://github.com/zulip/zulip/pull/33086/files#r1924602800) in the comments should be done in other places too.
- Added a comment in frontend code explaining the reason of not putting the same functionality on frontend markdown processor.

timabbott commented:
I think the strategy for this is correct, but the backend markdown processor changs are rather complex; tagging this as not for 10.0, since the issue is minor.
